sdk/src/asset_handlers/riff_io.rs

```diff
@@ -0,0 +1,424 @@
+// Copyright 2022 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::{
+    fs::{File, OpenOptions},
+    io::{Cursor, Seek, SeekFrom, Write},
+    path::Path,
+};
+
+use conv::ValueFrom;
+use riff::*;
+
+use crate::{
+    asset_io::{
+        AssetIO, AssetPatch, CAIRead, CAIReader, HashBlockObjectType, HashObjectPositions,
+        RemoteRefEmbed,
+    },
+    error::{Error, Result},
+    jumbf_io::get_file_extension,
+};
+
+static SUPPORTED_TYPES: [&str; 9] = [
+    "avi",
+    "wav",
+    "webp",
+    "image/webp",
+    "audio/x-wav",
+    "application/x-troff-msvideo",
+    "video/avi",
+    "video/msvideo",
+    "video/x-msvideo",
+];
+
+pub struct RiffIO {
+    #[allow(dead_code)]
+    riff_format: String, // can be used for specialized RIFF cases
+}
+
+const C2PA_CHUNK_ID: ChunkId = ChunkId {
+    value: [0x43, 0x32, 0x50, 0x41],
+}; // C2PA
+
+fn read_items<T>(iter: &mut T) -> Vec<T::Item>
+where
+    T: Iterator,
+{
+    let mut vec: Vec<T::Item> = Vec::new();
+    for item in iter {
+        vec.push(item);
+    }
+    vec
+}
+
+fn inject_c2pa<T>(chunk: &Chunk, file: &mut T, data: &[u8], format: &str) -> Result<ChunkContents>
+where
+    T: std::io::Seek + std::io::Read,
+{
+    let id = chunk.id();
+    let is_riff_chunk: bool = id == riff::RIFF_ID;
+
+    if is_riff_chunk || id == riff::LIST_ID {
+        let chunk_type = chunk.read_type(file).map_err(|_| {
+            Error::InvalidAsset("RIFF handler could not parse file format {format}".to_string())
+        })?;
+        let mut children = read_items(&mut chunk.iter(file));
+        let mut children_contents: Vec<ChunkContents> = Vec::new();
+
+        if is_riff_chunk {
+            // remove c2pa manifest store in RIFF chunk
+            children.retain(|c| c.id() != C2PA_CHUNK_ID);
+        }
+
+        // for non webp we can place at the front
+        // add c2pa manifest
+        if is_riff_chunk && !data.is_empty() && !format.contains("webp") {
+            children_contents.push(ChunkContents::Data(C2PA_CHUNK_ID, data.to_vec()));
+        }
+
+        for child in children {
+            children_contents.push(inject_c2pa(&child, file, data, format)?);
+        }
+
+        // for non webp we can place at the front
```

Comment:
Comment looks like a copy/paste error. For webp, the c2pa needs to be at the end, I take it?